### PR TITLE
Github App Tech Refactor

### PR DESCRIPTION
**Pull request by sasi2312**
Complete PR of App tech ticket

1. Once deployed in prod..pls add webhook secret in webhooks
2. Add that secret key in .env file named "WEBHOOK_SECRET"

### CODE REVIEW

**Comment by KDwevedi** [comment]
@sasi2312  Can you add a list of changes you're pushing in this PR?
- implemented A
- fixed B

Like so ^

**Comment by sasi2312** [comment]
These are the changes made on this PR, @KDwevedi 

- Created seperate handler class for each webhook events
- Implemented secretkey verifications in webhooks
- Created dispatcher file to call respective eventhandlers
- Created new handlers for issuecomments , pull requests
- Implemented webhook event & dispatcher as generic manner
- Craeted DB adapter using singleton pattern 
- Implemented the initialization of db adapter class as singleton instance
- Craeted Github adapter & clustering github related functions into this class
- Made changes on Ingestor file as per prod file
- Refactored some existing function in generic way

File: app.py [hunk 1]
```diff
@@ -27,49 +30,45 @@
...
     async with aiohttp.ClientSession() as session:
-        async with session.post(github_url_for_access_token, data=data, headers=headers) as response:
-            r = await response.json()
-            auth_token = (r)["access_token"]
-            headers = {
-                "Authorization": f"Bearer {auth_token}"
+        r = await GithubAdapter.get_github_data(code)
+
+        auth_token = (r)["access_token"]
+        headers = {
```
Comment (KDwevedi) [inline app.py:44]: if `headers` is being re-defined here, let's get rid of the definition on on Line#36

File: app.py [hunk 2]
```diff
@@ -27,49 +30,45 @@
...
         "Accept": "application/json"
     }
 
     async with aiohttp.ClientSession() as session:
-        async with session.post(github_url_for_access_token, data=data, headers=headers) as response:
-            r = await response.json()
-            auth_token = (r)["access_token"]
-            headers = {
-                "Authorization": f"Bearer {auth_token}"
+        r = await GithubAdapter.get_github_data(code)
```
Comment (KDwevedi) [inline app.py:41]: Let's name this variable better
Something like `responseData`

File: app.py [hunk 3]
```diff
@@ -27,49 +30,45 @@
...
+
+        auth_token = (r)["access_token"]
+        headers = {
+            "Authorization": f"Bearer {auth_token}"
+        }
+        async with session.get("https://api.github.com/user", headers=headers) as user_response:
```
Comment (KDwevedi) [inline app.py:47]: Let's wrap this in a function in Github Adaptor? `GetUser` ?

File: app.py [hunk 4]
```diff
@@ -81,24 +80,22 @@ async def comment_cleaner():
                 issue_id = comment["issue_id"]
                 comment = await TicketFeedbackHandler().deleteComment(owner, repo, comment_id)
                 print(f"Print Delete Task,{comment}", file=sys.stderr)
-                print(SupabaseInterface().deleteComment(issue_id))
+                print(SupabaseInterface.get_instance().deleteComment(issue_id))
 
 async def fetch_github_issues_from_repo(owner, repo):
-    url = f"https://api.github.com/repos/{owner}/{repo}/issues"
+    try:
+        response = await GithubAdapter.fetch_github_issues_from_repo(owner,repo)
```
Comment (KDwevedi) [inline app.py:87]: Let's just use camelCase for function definitions as well.

File: app.py [hunk 5]
```diff
@@ -81,24 +80,22 @@ async def comment_cleaner():
...
+        if response:
+            return response
+        else:
+            print(f"Failed to get issues: {response}")
+                
+    except Exception as e:
+        logger.info(e)
+        pass
```
Comment (KDwevedi) [inline app.py:95]: why is 
```python
pass
```

here?

File: handlers/installation_handler.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+from handlers.EventHandler import EventHandler
+
+class InstallationHandler(EventHandler):
+    async def handle_event(self, data, supabase_client):
+        # Implement your logic for handling issue events here
+        try:            
+            if data.get("installation") and data["installation"].get("account"):
+                # if data["action"] not in ["deleted", "suspend"]:
+                await TicketEventHandler().updateInstallation(data.get("installation"))
+            pass
```
Comment (KDwevedi) [inline handlers/installation_handler.py:12]: Why use `pass` here

Comment (sasi2312) [reply handlers/installation_handler.py:12]: Its just a placeholder here.

File: app.py (hunk 5 above)
Comment (sasi2312) [reply app.py:95]: Im not sure to raise a exception in this place ot not so i used pass here @KDwevedi

File: app.py (hunk 3 above)
Comment (sasi2312) [reply app.py:47]: okay ... will change

File: handlers/pull_request_handler.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                "ticket_url": data['pull_request']['issue_url'],
+                "ticket_complexity": "medium"
+            }
+            
+            points = supabase_client.get_data("url","ccbp_tickets",data['pull_request']['issue_url'],None)  
+            pr_data['points'] = points[0]['ticket_points'] if points else 0
+            
+            save_data = supabase_client.add_data(pr_data,"pr_history")            
+           
+            pass
```
Comment (KDwevedi) [inline handlers/pull_request_handler.py:32]: Why use `pass` here?

File: supabasedatapipeline/github_profile_render/ingestor.py [hunk 8]
```diff
@@ -65,11 +67,166 @@ def update(self, data):
...
+
+            textDimensions = cv2.getTextSize(githubUsername, cv2.FONT_HERSHEY_DUPLEX, 1.5, 2)
+            usernameWidth = textDimensions[0][0]
+            xCenter = 550
+
+            if githubUsername == "Shruti3004":
+                return None,None
+
+            #GET TEMPLATES
+            if milestone == 1:
```
Comment (KDwevedi) [inline supabasedatapipeline/github_profile_render/ingestor.py:125]: This should be a check using truth values of badge related table columns

Comment (sasi2312) [reply supabasedatapipeline/github_profile_render/ingestor.py:125]: Im not clear with this comment @KDwevedi  .... pls provide detailed explanations

File: utils/github_adapter.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+            headers = {
+                "Accept": "application/json"
+            }
+
+            async with aiohttp.ClientSession() as session:
+                async with session.post(github_url_for_access_token, data=data, headers=headers) as response:
+                    r = await response.json()                    
+                    return r
+                
+        except Exception as e:
```
Comment (KDwevedi) [inline utils/github_adapter.py:30]: Have we thought through exception handling and logging?
If, for example, this endpoint starts failing, we'll stop receiving github details of individuals registering through the discord flow.

We should probably raise an alert for that, for example on a discord webhook on the testing server/samagra server?

File: utils/github_adapter.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                    else:
+                        print(f"Failed to get issues: {response.status}")
+                        return None
+                        
+        except Exception as e:
+            logger.info(e)
+            return None
+            
+        
+    async def get_calssroom_data(assignment_id,page):
```
Comment (KDwevedi) [inline utils/github_adapter.py:68]: typo in function name
calssroom -> classroom

File: handlers/issue_comment_handler.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+                'id':data['comment']['id'],
+                'created_at':str(datetime.now()),
+                'updated_at':str(datetime.now())
+                
+            }
+            
+            save_data = supabase_client.add_data(comment_data,"ticket_comments")
+                     
+        except Exception as e:
+            logger.info(f"{datetime.now()}---jobs works")
```
Comment (KDwevedi) [inline handlers/issue_comment_handler.py:29]: we shouldn't be logging a success message if the function failed to save data to supabase

File: utils/github_adapter.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                if response.status == 200:
+                    print('Comment updated successfully.')
+                    return await response.json()
+                else:
+                    print(f'Error updating comment. Status code: {response.status}', file=sys.stderr)
+                    response_text = await response.text()
+                    print(f'Response body: {response_text}',file=sys.stderr)
+    
+    
+    async def deleteComment(self, owner, repo, comment_id):
```
Comment (KDwevedi) [inline utils/github_adapter.py:138]: let's have consistent naming for all variables and functions adhering to the [style guide](https://peps.python.org/pep-0008/#function-and-variable-names)

**Review by KDwevedi — CHANGES_REQUESTED** [review]
- the .get_instance() paradigm in SupabaseInterface will slowdown API interactions with database trying to instantiate and reinstantiate class instances. Let's simply convert that to being a module and import functions for use.
- Let's define exception handling better and introduce some observability.
- Let's not use `pass` as placeholder statements for anything. Many uses either have no functionality here or hide the fact that a crucial function has failed.
- Github Profile Rendering logic seems outdated.
  - It's referencing template names with old badge names (enthusiast/rising star)

File: utils/github_adapter.py (hunk 9 above)
Comment (sasi2312) [reply utils/github_adapter.py:30]: Im not clear about this comment... @KDwevedi  For now i raise an exception from except block

**Comment by sasi2312** [comment]
> * the .get_instance() paradigm in SupabaseInterface will slowdown API interactions with database trying to instantiate and reinstantiate class instances. Let's simply convert that to being a module and import functions for use.
> * Let's define exception handling better and introduce some observability.
> * Let's not use `pass` as placeholder statements for anything. Many uses either have no functionality here or hide the fact that a crucial function has failed.
> * Github Profile Rendering logic seems outdated.
>   
>   * It's referencing template names with old badge names (enthusiast/rising star)

**Comment by sasi2312** [comment]
- [ Github Profile Rendering logic seems outdated.
It's referencing template names with old badge names (enthusiast/rising star)]
 @KDwevedi 

- Can u please give me the context  of above statement

**Comment by KDwevedi** [comment]
> Actually the db adapter part is designed as per singleton pattern ... It returns single(same) instance for every call.Its not slow down any database related part

You're correct.

the singleton pattern is only 1.25 times slower here:
```python
import time

# Using a singleton class
class MySingletonClass:
    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(MySingletonClass, cls).__new__(cls, *args, **kwargs)
        return cls._instance

    def my_method(self):
        pass

# Timing the singleton class method calls
singleton_instance = MySingletonClass()
start = time.time()
for _ in range(1000000):
    singleton_instance.my_method()
end = time.time()
print(f"Singleton class method call time: {end - start} seconds")

# Using a module
def my_function():
    pass

# Timing the module function calls
start = time.time()
for _ in range(1000000):
    my_function()
end = time.time()
print(f"Module function call time: {end - start} seconds")
```
```bash
Singleton class method call time: 0.05184793472290039 seconds
Module function call time: 0.039575815200805664 seconds
```

While the class initialisation is 2-3x slower:
```python
import time

# Using a class
class MyClass:
    def my_method(self):
        pass

start = time.time()
for _ in range(1000000):
    instance = MyClass()
    instance.my_method()
end = time.time()
print(f"Class method call time: {end - start} seconds")

# Using a module
def my_function():
    pass

start = time.time()
for _ in range(1000000):
    my_function()
end = time.time()
print(f"Module function call time: {end - start} seconds")
```
```bash
Class method call time: 0.11432194709777832 seconds
Module function call time: 0.04015994071960449 seconds
```

I still feel like the singleton pattern is an unnecessary overhead, but @VedantKhairnar can decide if we want to migrate to a module.

Let's leave this be.

**Comment by KDwevedi** [comment]
> Can u please give me the context of above statement

The latest community badges are *Bronze*, *Silver*, *Gold*, *Ruby*, and *Diamond* according to the [leaderboard](https://www.codeforgovtech.in/leaderboard/)

The templates for the *Bronze*, *Silver*, and *Gold* badges are stored in the  [github_profile_assets](https://supabase.com/dashboard/project/kcavhjwafgtoqkqbbqrd/storage/buckets/github_profile_assets) bucket in supabase storage. 

The badges for *Ruby*, and *Diamond* can be made by copying the *Gold* or *Silver* or *Bronze* slide [here](https://docs.google.com/presentation/d/1-bNYTyKrilnZ4I8PVub6MGhNsDmTI9sN) 
(I've given edit access to sasikumar.s@cogniverselabs.com) and using the [Ruby](https://github.com/Code4GovTech/dmp-documentation/tree/main/assets/certificate_assets/Ruby.png) and [Diamond](https://github.com/Code4GovTech/dmp-documentation/tree/main/assets/certificate_assets/Diamond.png) png files.

In the new logic for generating github profile images you should
- [ ] Get rid of references to deprecated badges like enthusiast/rising star
- [ ] update the logic to also handle milestone 4 (Ruby) and milestone 5 (Diamond), as per point thresholds described in the community leaderboard
- [ ] add templates for these two badges
- [ ] validate the github_profile_data materialized view in supabase to make sure that milestone 4 and 5 are being generated correctly for valid point totals



Python script to generate the badges for reference:
```python
def generateQR(border_size=1, ghUsername=''):
    qr = segno.make_qr(f"https://github-app.c4gt.samagra.io/verify/{ghUsername}")
    buffer = BytesIO()
    # Save the QR code to a buffer, adjusting the border size here
    qr.save(buffer, kind='png', scale=1, border=border_size)
    qrBytes = buffer.getvalue()
    return qrBytes

profileData = client.table("github_profile_data").select("*").execute()
count = 0

for data in profileData.data:
    if data["dpg_points"]+data["classroom_points"]<10:
        continue

    

    #SET DATA
    milestone = data["milestone"]
    prsRaised = data["prs_raised"]
    prsReviewed = data["prs_reviewed"]
    prsMerged = data["prs_merged"]
    dpgPoints = data["dpg_points"]+data["classroom_points"]
    githubUsername = data["github_username"]

    githubLogoBytes = client.storage.from_('github_profile_assets').download("GHLogoSS.png")
    githubLogo_np = np.frombuffer(githubLogoBytes, np.uint8)
    githubLogo = cv2.imdecode(githubLogo_np, cv2.IMREAD_COLOR)

    qrBytes = generateQR(1, githubUsername)
    qr_np = np.frombuffer(qrBytes, np.uint8)
    qr =  cv2.imdecode(qr_np, cv2.IMREAD_COLOR)

    textDimensions = cv2.getTextSize(githubUsername, cv2.FONT_HERSHEY_DUPLEX, 1.5, 2)
    usernameWidth = textDimensions[0][0]
    xCenter = 550

    #GET TEMPLATES
    if milestone == 1:
        badgeName = 'Bronze'
    elif milestone == 2:
        badgeName = 'Silver'
    elif milestone == 3:
        badgeName = 'Gold'
    else:
        # Extend this logic for ruby and diamond
        print("WEIRD MILESTONE",milestone)
        
    displayTemplateName = f'Github Profile Template {badgeName} Display.png'
    verifiedTemplateName = f'Github Profile Template {badgeName} Verified.png'


    displayTemplateBytes = client.storage.from_('github_profile_assets').download(displayTemplateName)
    displayTemplate_np = np.frombuffer(displayTemplateBytes, np.uint8)
    displayTemplate = cv2.imdecode(displayTemplate_np, cv2.IMREAD_COLOR)

    width, height = displayTemplate.shape[:2]

    MILESTONES_COUNT_COORDINATES = (800, height - 673)
    PRS_RAISED_COORDINATES = (800, height - 608)
    PRS_REVIEWED_COORDINATES = (800, height - 546)
    PRS_MERGED_COORDINATES = (800, height - 483)
    DPG_POINTS_COORDINATES = (800, height - 418)
    QR_COORDINATES = (855, height - 620)
    GITHUB_USERNAME = (xCenter - usernameWidth//2, height - 792)
    GITHUB_LOGO_COORDINATES = (GITHUB_USERNAME[0] - 90, height - 850)

    ghScaleFactor = 0.08
    newGHSize = (int(displayTemplate.shape[1] * ghScaleFactor), int(displayTemplate.shape[1] * ghScaleFactor))
    ghResized = cv2.resize(githubLogo, newGHSize, interpolation=cv2.INTER_AREA)
    displayTemplate[GITHUB_LOGO_COORDINATES[1]:GITHUB_LOGO_COORDINATES[1]+ghResized.shape[0], GITHUB_LOGO_COORDINATES[0]:GITHUB_LOGO_COORDINATES[0]+ghResized.shape[1]] = ghResized

    scaleFactor = 0.107
    newSize = (int(displayTemplate.shape[1] * scaleFactor), int(displayTemplate.shape[1] * scaleFactor))
    qrResized = cv2.resize(qr, newSize, interpolation=cv2.INTER_AREA)
    displayTemplate[QR_COORDINATES[1]:QR_COORDINATES[1]+qrResized.shape[0], QR_COORDINATES[0]:QR_COORDINATES[0]+qrResized.shape[1]] = qrResized



    font = cv2.FONT_HERSHEY_SIMPLEX
    font_size = 1.0
    color = (255, 255, 255)
    thickness = 2

    cv2.putText(displayTemplate, str(milestone), MILESTONES_COUNT_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsRaised), PRS_RAISED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsReviewed), PRS_REVIEWED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(prsMerged), PRS_MERGED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, str(dpgPoints), DPG_POINTS_COORDINATES, font, font_size, color, thickness)
    cv2.putText(displayTemplate, f"@{githubUsername}", GITHUB_USERNAME, cv2.FONT_HERSHEY_DUPLEX, 1.5, (216,216,216), thickness)

    verifiedTemplateBytes = client.storage.from_('github_profile_assets').download(verifiedTemplateName)
    verifiedTemplate_np = np.frombuffer(verifiedTemplateBytes, np.uint8)
    verifiedTemplate = cv2.imdecode(verifiedTemplate_np, cv2.IMREAD_COLOR)

    width, height = verifiedTemplate.shape[:2]

    MILESTONES_COUNT_COORDINATES = (800, height - 673)
    PRS_RAISED_COORDINATES = (800, height - 608)
    PRS_REVIEWED_COORDINATES = (800, height - 546)
    PRS_MERGED_COORDINATES = (800, height - 483)
    DPG_POINTS_COORDINATES = (800, height - 418)
    QR_COORDINATES = (855, height - 620)
    GITHUB_USERNAME = (xCenter - usernameWidth//2, height - 792)
    GITHUB_LOGO_COORDINATES = (GITHUB_USERNAME[0] - 90, height - 850)

    ghScaleFactor = 0.08
    newGHSize = (int(verifiedTemplate.shape[1] * ghScaleFactor), int(verifiedTemplate.shape[1] * ghScaleFactor))
    ghResized = cv2.resize(githubLogo, newGHSize, interpolation=cv2.INTER_AREA)
    verifiedTemplate[GITHUB_LOGO_COORDINATES[1]:GITHUB_LOGO_COORDINATES[1]+ghResized.shape[0], GITHUB_LOGO_COORDINATES[0]:GITHUB_LOGO_COORDINATES[0]+ghResized.shape[1]] = ghResized

    scaleFactor = 0.107
    newSize = (int(verifiedTemplate.shape[1] * scaleFactor), int(verifiedTemplate.shape[1] * scaleFactor))
    qrResized = cv2.resize(qr, newSize, interpolation=cv2.INTER_AREA)
    verifiedTemplate[QR_COORDINATES[1]:QR_COORDINATES[1]+qrResized.shape[0], QR_COORDINATES[0]:QR_COORDINATES[0]+qrResized.shape[1]] = qrResized



    font = cv2.FONT_HERSHEY_SIMPLEX
    font_size = 1.0
    color = (255, 255, 255)
    thickness = 2

    cv2.putText(verifiedTemplate, str(milestone), MILESTONES_COUNT_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsRaised), PRS_RAISED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsReviewed), PRS_REVIEWED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(prsMerged), PRS_MERGED_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, str(dpgPoints), DPG_POINTS_COORDINATES, font, font_size, color, thickness)
    cv2.putText(verifiedTemplate, f"@{githubUsername}", GITHUB_USERNAME, cv2.FONT_HERSHEY_DUPLEX, 1.5, (216,216,216), thickness)

    display = cv2.imencode(".jpg", displayTemplate)[1].tobytes()
    verified = cv2.imencode(".jpg", verifiedTemplate)[1].tobytes()
    displayFilename = f'{data["discord_id"]}-c4gt-contributions.jpeg'
    verifiedFileName = f'{githubUsername}-c4gt-contributions.jpeg'
    try:
        client.storage.from_("c4gt-github-profile").remove(displayFilename)
        client.storage.from_("c4gt-github-profile").remove(verifiedFileName)
    except Exception:
        pass

    # self.supabase.client.storage.from_(self.storageBucket).remove(displayFilename)
    client.storage.from_("c4gt-github-profile").upload(displayFilename, display, {"content-type": "image/jpeg"})
    client.storage.from_("c4gt-github-profile").upload(verifiedFileName, verified, {"content-type": "image/jpeg"})
```

**Comment by KDwevedi** [comment]
Also @sasi2312, I don't see any commits post review, I hope the rest of the feedback on extraneous `pass` statements and more robust error handling can be incorporated.

**Pull request merged**